### ISSUE TITLE
Include end bound in `pv.voxelize()` and `pv.voxelize_volume()` to avoid cropping the mesh

### ISSUE THREAD

**Issue by GuillaumeBroggi** [pyvista/pyvista#6609]
### Describe the feature you would like to be added.

Voxelizing a mesh with `pv.voxelize()` or `pv.voxelize_volume()` will crop the mesh. 

For instance, for a simple cube with the default parameters, the mesh bounds are `[-0.5, 0.5]`: 

```python
import pyvista as pv
cube = pv.Cube()
cube
```
<table style='width: 100%;'><tr><th>Header</th><th>Data Arrays</th></tr><tr><td>
<table style='width: 100%;'>
<tr><th>PolyData</th><th>Information</th></tr>
<tr><td>N Cells</td><td>6</td></tr>
<tr><td>N Points</td><td>8</td></tr>
<tr><td>N Strips</td><td>0</td></tr>
<tr><td>X Bounds</td><td>-5.000e-01, 5.000e-01</td></tr>
<tr><td>Y Bounds</td><td>-5.000e-01, 5.000e-01</td></tr>
<tr><td>Z Bounds</td><td>-5.000e-01, 5.000e-01</td></tr>
<tr><td>N Arrays</td><td>3</td></tr>
</table>

</td><td>
<table style='width: 100%;'>
<tr><th>Name</th><th>Field</th><th>Type</th><th>N Comp</th><th>Min</th><th>Max</th></tr>
<tr><td>Normals</td><td>Points</td><td>float32</td><td>3</td><td>-1.000e+00</td><td>1.000e+00</td></tr>
<tr><td>TCoords</td><td>Points</td><td>float32</td><td>2</td><td>-1.000e+00</td><td>1.000e+00</td></tr>
<tr><td>FaceIndex</td><td>Cells</td><td>int64</td><td>1</td><td>0.000e+00</td><td>5.000e+00</td></tr>
</table>

</td></tr> </table>

After voxelization, the mesh bounds are `[-0.5, 0.3]`, the missing `0.2` corresponding to a missing row of cells: 

```python
pv.voxelize(cube, density=0.2)
```
<table style='width: 100%;'><tr><th>Header</th><th>Data Arrays</th></tr><tr><td>
<table style='width: 100%;'>
<tr><th>UnstructuredGrid</th><th>Information</th></tr>
<tr><td>N Cells</td><td>64</td></tr>
<tr><td>N Points</td><td>125</td></tr>
<tr><td>X Bounds</td><td>-5.000e-01, 3.000e-01</td></tr>
<tr><td>Y Bounds</td><td>-5.000e-01, 3.000e-01</td></tr>
<tr><td>Z Bounds</td><td>-5.000e-01, 3.000e-01</td></tr>
<tr><td>N Arrays</td><td>2</td></tr>
</table>

</td><td>
<table style='width: 100%;'>
<tr><th>Name</th><th>Field</th><th>Type</th><th>N Comp</th><th>Min</th><th>Max</th></tr>
<tr><td>vtkOriginalPointIds</td><td>Points</td><td>int64</td><td>1</td><td>0.000e+00</td><td>1.240e+02</td></tr>
<tr><td>vtkOriginalCellIds</td><td>Cells</td><td>int64</td><td>1</td><td>0.000e+00</td><td>6.300e+01</td></tr>
</table>

</td></tr> </table>

One could expect the cube mesh to remain whole. Even if cropping is the intended behavior, there is no straightforward method to obtain a complete voxelized mesh of the cube.

As mentioned in https://github.com/pyvista/pyvista/issues/4312#issuecomment-1516994440, using `np.linspace()` instead of `np.arange()` would implement the proposed feature. Would a PR implementing this be welcome, or would it bring unwanted side effects?

**Note**

https://github.com/pyvista/pyvista/pull/6514 does not implement this behavior, instead the mesh is padded to enclose the cube:

```python
pv.voxelize(cube, density=0.2, enclosed=True)
```
<table style='width: 100%;'><tr><th>Header</th><th>Data Arrays</th></tr><tr><td>
<table style='width: 100%;'>
<tr><th>UnstructuredGrid</th><th>Information</th></tr>
<tr><td>N Cells</td><td>216</td></tr>
<tr><td>N Points</td><td>343</td></tr>
<tr><td>X Bounds</td><td>-6.000e-01, 6.000e-01</td></tr>
<tr><td>Y Bounds</td><td>-6.000e-01, 6.000e-01</td></tr>
<tr><td>Z Bounds</td><td>-6.000e-01, 6.000e-01</td></tr>
<tr><td>N Arrays</td><td>2</td></tr>
</table>

</td><td>
<table style='width: 100%;'>
<tr><th>Name</th><th>Field</th><th>Type</th><th>N Comp</th><th>Min</th><th>Max</th></tr>
<tr><td>vtkOriginalPointIds</td><td>Points</td><td>int64</td><td>1</td><td>0.000e+00</td><td>3.420e+02</td></tr>
<tr><td>vtkOriginalCellIds</td><td>Cells</td><td>int64</td><td>1</td><td>0.000e+00</td><td>2.150e+02</td></tr>
</table>

</td></tr> </table>

**Comment by tkoyama010** [comment]
Thank you for your suggestion. How about add it as an new option first? If the new method is better, we can make anothor breaking-change and make it the default. Pull requests are always welcome.

**Comment by GuillaumeBroggi** [comment]
In that case, I can start with the following signature that will not break anything:
```python
voxelize(mesh, density=None, check_surface=True, enclosed=False, include_bounds=False)
```

Maybe it could be made into a more generic function in the future, something like (just writing down some ideas): 

```python
voxelize(mesh, density=None, check_surface=True, range=["include", "exclude"])
```
where the behavior are defined for each bound:
- `float`:`: explicit value for the bound
- `"exclude"`: exclude the bound 
- `"include"`: include the bound
- `"enclosed"`: pad the mesh to ensure the bound is enclosed
 
The current behavior would be achieved with `range=["include", "exclude"]`. Using range=[float, float]` or `range=["include", "include"]` would update the `density` if required to accommodate both bounds.

**Comment by tkoyama010** [comment]
Awesome. Let's try it!

**Comment by GuillaumeBroggi** [comment]
Implemented in both #6610 and #6611
